File: Exec Time/exec time enhanced/enhanced_exec_time.py
# ExecutionTime = ClockCycleTime * Σ(InstructionCount_i * CPI_i)
# Input: clock cycle time, instruction counts and CPIs for N instruction types
# Output: execution time in seconds


def get_positive_float(prompt):
    """Get a positive float from user with validation."""
    while True:
        try:
            value = float(input(prompt))
            if value > 0:
                return value
            else:
                print("   Please enter a positive number.")
        except ValueError:
            print("   Invalid input. Please enter a valid number.")


def get_non_negative_int(prompt):
    """Get a non-negative integer from user with validation."""
    while True:
        try:
            value = int(input(prompt))
            if value >= 0:
                return value
            else:
                print("   Please enter a non-negative integer.")
        except ValueError:
            print("   Invalid input. Please enter a whole number.")


def get_instruction_metrics(num_types):
    """Collect instruction counts and CPIs for all instruction types."""
    counts = []
    cpis = []
    
    for i in range(num_types):
        count = get_non_negative_int(
            f"{2 + i * 2}) The counts of Type {i + 1} instruction (Instruction{i + 1}_count): "
        )
        counts.append(count)
        
        cpi = get_positive_float(
            f"{3 + i * 2}) The CPI of Type {i + 1} instruction (CPI_{i + 1}): "
        )
        cpis.append(cpi)
    
    return counts, cpis


def calculate_execution_time(clock_time, counts, cpis):
    """Calculate total execution time."""
    total_cycles = sum(count * cpi for count, cpi in zip(counts, cpis))
    return clock_time * total_cycles


def main():
    """Calculate and display execution time based on instruction metrics."""
    print("CSC1104 – Execution Time Calculator\n")

    # Ask user how many instruction types
    while True:
        try:
            num_types = int(input("Enter the number of instruction types: "))
            if num_types > 0:
                break
            else:
                print("   Please enter a positive integer.")
        except ValueError:
            print("   Invalid input. Please enter a whole number.")

    # Get clock cycle time
    t_clk = get_positive_float("1) The value of clock cycle time (in second): ")

    # Get instruction counts and CPIs
    counts, cpis = get_instruction_metrics(num_types)

    # Calculate execution time
    exec_time = calculate_execution_time(t_clk, counts, cpis)

    # Display result
    print("\n-------------------------------------------")
    print(f"The execution time of this software program is {exec_time:.6f} second.")
    print("-------------------------------------------")


if __name__ == "__main__":
    main()

File: Exec Time/exec time enhanced/enhanced_exec_time.c
#include <stdio.h>
#include <stdlib.h>  // for malloc()

// ExecutionTime = ClockCycleTime * Σ(InstructionCount_i * CPI_i)
// Input: clock cycle time, instruction counts and CPIs for user-defined instruction types
// Output: execution time in seconds

int main(void) {
    double t_clk;
    int num_types;

    // Helper macro to clear invalid input buffer, clears garbage data
    #define CLEAR_INPUT() while (getchar() != '\n')

    // Get number of instruction types
    while (1) {
        printf("Enter the number of instruction types: ");
        if (scanf("%d", &num_types) == 1 && num_types > 0)
            break;
        printf("   Invalid input. Please enter a positive whole number.\n");
        CLEAR_INPUT();
    }

    // Dynamically allocate arrays
    long long *count = malloc(num_types * sizeof(long long));
    double *cpi = malloc(num_types * sizeof(double));

    if (count == NULL || cpi == NULL) {
        printf("Memory allocation failed.\n");
        return 1;
    }

    // Get clock cycle time
    while (1) {
        printf("1) The value of clock cycle time (in second): ");
        if (scanf("%lf", &t_clk) == 1 && t_clk > 0)
            break;
        printf("   Invalid input. Please enter a positive number.\n");
        CLEAR_INPUT();
    }

    // Get counts and CPIs for each instruction type using for-loop
    for (int i = 0; i < num_types; i++) {
        // Instruction count
        while (1) {
            printf("%d) The counts of Type %d instruction (Instruction%d_count): ",
                   2 + i * 2, i + 1, i + 1);
            if (scanf("%lld", &count[i]) == 1 && count[i] >= 0)
                break;
            printf("   Invalid input. Please enter a non-negative whole number.\n");
            CLEAR_INPUT();
        }

        // CPI
        while (1) {
            printf("%d) The CPI of Type %d instruction (CPI_%d): ",
                   3 + i * 2, i + 1, i + 1);
            if (scanf("%lf", &cpi[i]) == 1 && cpi[i] > 0)
                break;
            printf("   Invalid input. Please enter a positive number.\n");
            CLEAR_INPUT();
        }
    }

    // Compute total cycles
    double total_cycles = 0;
    for (int i = 0; i < num_types; i++) {
        total_cycles += count[i] * cpi[i];
    }

    // Compute execution time
    double exec_time = t_clk * total_cycles;

    // Display result
    printf("\n-------------------------------------------\n");
    printf("The execution time of this software program is %.6f seconds.\n", exec_time);
    printf("-------------------------------------------\n");

    // Free dynamically allocated memory
    free(count);
    free(cpi);

    return 0;
}
